File: codec_selector.py
"""
CodecDB-inspired data-driven encoding selection for Parquet columns.

Extracts per-column features (cardinality ratio, sortedness, value type/range)
and selects per-column encoding parameters using rules derived from the
CodecDB paper (Jiang et al., SIGMOD '21).

Reference:
  Hao Jiang, Chunwei Liu, John Paparrizos, Andrew A. Chien, Jihong Ma,
  Aaron J. Elmore. "Good to the Last Bit: Data-Driven Encoding with CodecDB."
  SIGMOD '21. https://doi.org/10.1145/3448016.3457283
"""

from __future__ import annotations

from typing import Any, Dict

import numpy as np
import pyarrow as pa

CARDINALITY_THRESHOLD = 0.3


def compute_column_features(table: pa.Table) -> Dict[str, Dict[str, Any]]:
    """
    Extract encoding-relevant features for each column in *table*.

    Per-column features:
      dtype, n_rows, n_unique, cardinality_ratio, is_sorted,
      and type-specific extras (max/min for ints, mean_value_length for strings).
    """
    features: Dict[str, Dict[str, Any]] = {}
    n_rows = table.num_rows

    for name in table.column_names:
        col = table.column(name)
        arr = col.combine_chunks()

        feat: Dict[str, Any] = {
            "dtype": str(arr.type),
            "n_rows": n_rows,
        }

        try:
            n_unique = len(arr.unique())
        except Exception:
            n_unique = n_rows
        feat["n_unique"] = n_unique
        feat["cardinality_ratio"] = n_unique / n_rows if n_rows > 0 else 1.0

        is_numeric = pa.types.is_integer(arr.type) or pa.types.is_floating(arr.type)
        is_string = pa.types.is_string(arr.type) or pa.types.is_large_string(arr.type)

        if is_numeric and n_rows > 1:
            try:
                np_arr = arr.to_numpy(zero_copy_only=False)
                if pa.types.is_floating(arr.type):
                    valid = np_arr[~np.isnan(np_arr)]
                else:
                    valid = np_arr
                if len(valid) > 1:
                    feat["is_sorted"] = bool(np.all(np.diff(valid) >= 0))
                    feat["max_value"] = int(np.max(valid))
                    feat["min_value"] = int(np.min(valid))
                else:
                    feat["is_sorted"] = True
                    feat["max_value"] = int(valid[0]) if len(valid) else 0
                    feat["min_value"] = feat["max_value"]
            except Exception:
                feat["is_sorted"] = False
        elif is_string and n_rows > 1:
            try:
                py_list = arr.to_pylist()
                non_null = [s for s in py_list if s is not None]
                feat["is_sorted"] = all(a <= b for a, b in zip(non_null, non_null[1:]))
                lengths = [len(s) for s in non_null]
                feat["mean_value_length"] = (
                    sum(lengths) / len(lengths) if lengths else 0
                )
            except Exception:
                feat["is_sorted"] = False
        else:
            feat.setdefault("is_sorted", n_rows <= 1)

        features[name] = feat

    return features


def select_encoding_params(table: pa.Table) -> Dict[str, Any]:
    """
    Select per-column Parquet encoding parameters based on column features.

    Returns a dict with:
      use_dictionary  – list of column names that should use dictionary encoding
                        (passed directly to ``pq.write_table(use_dictionary=...)``)
      column_encoding – dict[column_name, str] | None  (per-column encoding override)
      features        – dict[column_name, feature_dict]

    Rules (CodecDB-inspired):
      * cardinality_ratio < CARDINALITY_THRESHOLD → dictionary (PyArrow default)
      * high-cardinality sorted integers           → DELTA_BINARY_PACKED
      * high-cardinality strings                   → DELTA_BYTE_ARRAY
      * otherwise                                  → PLAIN (no dictionary)
    """
    features = compute_column_features(table)

    dict_columns: list[str] = []
    column_encoding: Dict[str, str] = {}

    for name, feat in features.items():
        cr = feat.get("cardinality_ratio", 1.0)
        is_sorted = feat.get("is_sorted", False)
        dtype_str = feat.get("dtype", "")
        is_int = "int" in dtype_str
        is_str = "string" in dtype_str or "utf8" in dtype_str

        if cr < CARDINALITY_THRESHOLD:
            dict_columns.append(name)
        else:
            if is_int and is_sorted:
                column_encoding[name] = "DELTA_BINARY_PACKED"
            elif is_str:
                column_encoding[name] = "DELTA_BYTE_ARRAY"
            else:
                column_encoding[name] = "PLAIN"

    return {
        "use_dictionary": dict_columns,
        "column_encoding": column_encoding if column_encoding else None,
        "features": features,
    }

File: java-server/src/main/java/com/multimodal/mcp/codec/CodecSelector.java
package com.multimodal.mcp.codec;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

import org.apache.arrow.vector.FieldVector;
import org.apache.arrow.vector.VectorSchemaRoot;
import org.apache.arrow.vector.types.pojo.ArrowType;
import org.springframework.stereotype.Component;

/**
 * CodecDB-inspired data-driven encoding selection for Parquet columns.
 */
@Component
public class CodecSelector {

    public static final double CARDINALITY_THRESHOLD = 0.3;

    public Map<String, Map<String, Object>> computeColumnFeatures(VectorSchemaRoot table) {
        Map<String, Map<String, Object>> features = new HashMap<>();
        int nRows = table.getRowCount();

        for (FieldVector vector : table.getFieldVectors()) {
            String name = vector.getField().getName();
            Map<String, Object> feat = new HashMap<>();
            feat.put("dtype", vector.getField().getType().toString());
            feat.put("n_rows", nRows);

            int nUnique = estimateUniqueCount(vector, nRows);
            feat.put("n_unique", nUnique);
            feat.put("cardinality_ratio", nRows > 0 ? (double) nUnique / nRows : 1.0);

            ArrowType type = vector.getField().getType();
            boolean isNumeric = isIntegerType(type) || isFloatingType(type);
            boolean isString = isStringType(type);

            if (isNumeric && nRows > 1) {
                populateNumericFeatures(vector, nRows, feat);
            } else if (isString && nRows > 1) {
                populateStringFeatures(vector, nRows, feat);
            } else {
                feat.putIfAbsent("is_sorted", nRows <= 1);
            }

            features.put(name, feat);
        }
        return features;
    }

    public Map<String, Object> selectEncodingParams(VectorSchemaRoot table) {
        Map<String, Map<String, Object>> features = computeColumnFeatures(table);
        List<String> dictColumns = new ArrayList<>();
        Map<String, String> columnEncoding = new HashMap<>();

        for (Map.Entry<String, Map<String, Object>> entry : features.entrySet()) {
            String name = entry.getKey();
            Map<String, Object> feat = entry.getValue();
            double cr = toDouble(feat.get("cardinality_ratio"), 1.0);
            boolean isSorted = Boolean.TRUE.equals(feat.get("is_sorted"));
            String dtypeStr = String.valueOf(feat.get("dtype"));
            String dtypeLower = dtypeStr.toLowerCase();
            boolean isInt = dtypeLower.contains("int") && !dtypeLower.contains("point");
            boolean isStr = dtypeLower.contains("utf8") || dtypeLower.contains("string");

            if (cr < CARDINALITY_THRESHOLD) {
                dictColumns.add(name);
            } else {
                if (isInt && isSorted) {
                    columnEncoding.put(name, "DELTA_BINARY_PACKED");
                } else if (isStr) {
                    columnEncoding.put(name, "DELTA_BYTE_ARRAY");
                } else {
                    columnEncoding.put(name, "PLAIN");
                }
            }
        }

        Map<String, Object> result = new HashMap<>();
        result.put("use_dictionary", dictColumns);
        result.put("column_encoding", columnEncoding.isEmpty() ? null : columnEncoding);
        result.put("features", features);
        return result;
    }

    private void populateNumericFeatures(FieldVector vector, int nRows, Map<String, Object> feat) {
        try {
            List<Double> values = new ArrayList<>();
            for (int i = 0; i < nRows; i++) {
                if (!vector.isNull(i)) {
                    values.add(readNumericValue(vector, i));
                }
            }
            if (values.size() > 1) {
                boolean sorted = true;
                for (int i = 1; i < values.size(); i++) {
                    if (values.get(i - 1) > values.get(i)) {
                        sorted = false;
                        break;
                    }
                }
                feat.put("is_sorted", sorted);
                double max = values.stream().mapToDouble(Double::doubleValue).max().orElse(0);
                double min = values.stream().mapToDouble(Double::doubleValue).min().orElse(0);
                feat.put("max_value", (long) max);
                feat.put("min_value", (long) min);
            } else if (values.size() == 1) {
                feat.put("is_sorted", true);
                feat.put("max_value", values.get(0).longValue());
                feat.put("min_value", values.get(0).longValue());
            } else {
                feat.put("is_sorted", true);
                feat.put("max_value", 0L);
                feat.put("min_value", 0L);
            }
        } catch (Exception e) {
            feat.put("is_sorted", false);
        }
    }

    private void populateStringFeatures(FieldVector vector, int nRows, Map<String, Object> feat) {
        try {
            List<String> nonNull = new ArrayList<>();
            for (int i = 0; i < nRows; i++) {
                if (!vector.isNull(i)) {
                    Object value = vector.getObject(i);
                    if (value != null) {
                        nonNull.add(String.valueOf(value));
                    }
                }
            }
            boolean sorted = true;
            for (int i = 1; i < nonNull.size(); i++) {
                if (nonNull.get(i - 1).compareTo(nonNull.get(i)) > 0) {
                    sorted = false;
                    break;
                }
            }
            feat.put("is_sorted", sorted);
            if (!nonNull.isEmpty()) {
                int totalLen = nonNull.stream().mapToInt(String::length).sum();
                feat.put("mean_value_length", (double) totalLen / nonNull.size());
            } else {
                feat.put("mean_value_length", 0.0);
            }
        } catch (Exception e) {
            feat.put("is_sorted", false);
        }
    }

    private int estimateUniqueCount(FieldVector vector, int nRows) {
        try {
            Set<Object> unique = new HashSet<>();
            for (int i = 0; i < nRows; i++) {
                if (!vector.isNull(i)) {
                    unique.add(vector.getObject(i));
                }
            }
            return unique.size();
        } catch (Exception e) {
            return nRows;
        }
    }

    private double readNumericValue(FieldVector vector, int index) {
        Object value = vector.getObject(index);
        if (value instanceof Number number) {
            return number.doubleValue();
        }
        return Double.parseDouble(String.valueOf(value));
    }

    private static boolean isIntegerType(ArrowType type) {
        return type instanceof ArrowType.Int;
    }

    private static boolean isFloatingType(ArrowType type) {
        return type instanceof ArrowType.FloatingPoint;
    }

    private static boolean isStringType(ArrowType type) {
        return type instanceof ArrowType.Utf8 || type instanceof ArrowType.LargeUtf8;
    }

    private static double toDouble(Object value, double defaultValue) {
        if (value instanceof Number number) {
            return number.doubleValue();
        }
        if (value == null) {
            return defaultValue;
        }
        try {
            return Double.parseDouble(String.valueOf(value));
        } catch (NumberFormatException e) {
            return defaultValue;
        }
    }
}
